File: 85. Maximal Rectangle/85. Maximal Rectangle.go
func maximalRectangle(matrix [][]byte) int {
    if len(matrix) == 0 || len(matrix[0]) == 0 {
        return 0
    }
    rows, cols := len(matrix), len(matrix[0])
    heights := make([]int, cols)
    maxArea := 0

    for i := 0; i < rows; i++ {
        for j := 0; j < cols; j++ {
            if matrix[i][j] == '1' {
                heights[j]++
            } else {
                heights[j] = 0
            }
        }
        maxArea = max(maxArea, largestRectangleArea(heights))
    }

    return maxArea
}

func largestRectangleArea(heights []int) int {
    stack := []int{}
    heights = append(heights, 0)
    maxArea := 0

    for i := 0; i < len(heights); i++ {
        for len(stack) > 0 && heights[i] < heights[stack[len(stack)-1]] {
            height := heights[stack[len(stack)-1]]
            stack = stack[:len(stack)-1]
            width := i
            if len(stack) > 0 {
                width = i - stack[len(stack)-1] - 1
            }
            maxArea = max(maxArea, height*width)
        }
        stack = append(stack, i)
    }

    return maxArea
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: 85. Maximal Rectangle/85. Maximal Rectangle.rb
# @param {Character[][]} matrix
# @return {Integer}
def maximal_rectangle(matrix)
    return 0 if matrix.empty? || matrix[0].empty?
    rows, cols = matrix.length, matrix[0].length
    heights = Array.new(cols, 0)
    max_area = 0

    rows.times do |i|
        cols.times do |j|
            if matrix[i][j] == '1'
                heights[j] += 1
            else
                heights[j] = 0
            end
        end
        max_area = [max_area, largest_rectangle_area(heights)].max
    end

    max_area
end

def largest_rectangle_area(heights)
    stack = []
    heights << 0
    max_area = 0
    i = 0

    while i < heights.length
        if stack.empty? || heights[i] >= heights[stack.last]
            stack.push(i)
            i += 1
        else
            height = heights[stack.pop]
            width = stack.empty? ? i : i - stack.last - 1
            max_area = [max_area, height * width].max
        end
    end

    max_area
end